File: C/数字转汉字.c
#include <stdio.h>

static char *number[] = {
  "零","一","二","三","四","五","六","七","八","九"
};

static char *unit(int n)
{
  static char* unitList[] = {
    "十", "百", "千", "万", "十", "百", "千", "亿"
  };
  if(n<=0) return "";
  return unitList[ (n-1)%8 ];
}

void printHanzi(const char* buf)
{
  if(buf == NULL) return;
  if(*buf=='-')
  {
    printf("负");
    ++buf;
  }
  while(*buf == '0') ++buf; //去除头0
  int length = 0;
  while('0'<=buf[length] && buf[length]<='9') ++length;
  if(length == 0)
  {
    printf("%s",number[0]);
    return;
  }
  int i = 0;
  int zeroflag = -1, wanflag = 0;
  while(i<length)
  {
    int weight = length - 1 - i; //当前位置, 个位为0
    int c = buf[i] - '0';       //当前位的数
    if(c == 0)
    {
      if(!zeroflag) //处理 '0' 位
      {
        wanflag = ( (weight+1) %8 != 0); //设置 单位 '万' 的输出: 若当前位的上一位不是 '亿', 则输出接下来遇到的第一个 万
      }
      if(wanflag == 1 && (weight % 8 == 4)) // 若为万位 并符合输出条件, 输出 单位, 并设置标志位,不再输出'万'
      {
        printf("%s",unit(weight));
        wanflag = 0;
      }
      if(weight %8 == 0) //若为 亿 位, 输出 单位, 并设置标志位 不再输出 '万'
      {
        printf("%s",unit(weight));
        wanflag = 0;
      }
      zeroflag = 1;
    }
    else
    {
      if(zeroflag == 1) //输出 '零'
      {
        printf("%s",number[0]);
      }
      if(!( i == 0 && c == 1 && length%4 == 2 ) ) //判断是否以 十 开头 (十, 十万, 十亿, 不输出 '一')
      {
        if( c == 2 &&  weight % 4 != 1 && zeroflag != 0) // 判断是否 是 十, 十万, 十亿 不能输出 '两'
          printf("%s", "两");
        else
          printf("%s",number[c]);
      }
      printf("%s",unit(weight));
      zeroflag =  0;
    }
    ++i;
  }
}

File: Javascript/数字转汉字.js
function numToUpper(numStr) {
  numStr = numStr.toString()
  let number = new Array("零","一","二","三","四","五","六","七","八","九")
  let unitList = new Array("十", "百", "千", "万", "十", "百", "千", "亿")
  let neg = '负'
  let two = '两'
  let unit = (i)=> i>0 ? unitList[(i-1)%8] : ""
  var result = ''
  var index = 0, length = 0
  if(numStr.charAt(index) == '-') {
    result+=neg
    ++index
  }
  while(numStr.charAt(index) == '0') ++index
  numStr = numStr.substring(index)
  index = 0
  while(parseInt(numStr.charAt(length)) >= 0) ++length
  if(length ==0) {
    return number[0]
  }
  var i = 0, zeroflag = -1, wanflag = 0
  while(i < length) {
    let weight = length - 1 - i
    let c = parseInt(numStr.charAt(i))
    if(c == 0) {
      if(zeroflag == 0) {
        wanflag = (weight+1) %8 != 0
      }
      if(wanflag == 1 && weight % 8 == 4) {
        result += unit(weight)
        wanflag = 0
      }
      if(weight % 8 == 0) {
        result += unit(weight)
        wanflag = 0
      }
      zeroflag = 1
    } else {
      if(zeroflag == 1) {
        result += number[0]
      }
      if(!( i == 0 && c == 1 && length%4 == 2 )) {
        if( c == 2 && (weight % 4 == 0 && weight !=0 && zeroflag != 0 || weight % 4 > 1) ) {
          result += two
        } else {
          result += number[c]
        }
      }
      result += unit(weight)
      zeroflag = 0
    }
    ++i
  }
  return result
}


function numToBigUpper(numStr) {
  numStr = numStr.toString()
  let number = new Array("零","一","二","三","四","五","六","七","八","九")
  let unitList = new Array("十", "百", "千")
  let bigUnitList = new Array("万","亿","兆","京","垓","秭","壤","沟","涧","正","载")
  let neg = '负'
  let two = '两'
  let unit = (i)=> {
    if(i<=0) return ""
    i = (i-1) % ((unitList.length + 1) * bigUnitList.length) + 1
    let n = i % (unitList.length+1)
    return (n == 0) ? bigUnitList[ i / (unitList.length+1) - 1] : unitList[n - 1]
  }
  var result = ''
  var index = 0, length = 0
  if(numStr.charAt(index) == '-') {
    result+=neg
    ++index
  }
  while(numStr.charAt(index) == '0') ++index
  numStr = numStr.substring(index)
  index = 0
  while(parseInt(numStr.charAt(length)) >= 0) ++length
  if(length ==0) {
    return number[0]
  }
  var i = 0, zeroflag = -1, wanflag = 0
  while(i < length) {
    let weight = length - 1 - i
    let c = parseInt(numStr.charAt(i))
    if(c == 0) {
      if(zeroflag == 0) {
        wanflag = (weight+1) %8 != 0
      }
      if(wanflag == 1 && weight % 8 == 4) {
        result += unit(weight)
        wanflag = 0
      }
      if(weight % 8 == 0) {
        result += unit(weight)
        wanflag = 0
      }
      zeroflag = 1
    } else {
      if(zeroflag == 1) {
        result += number[0]
      }
      if(!( i == 0 && c == 1 && length%4 == 2 )) {
        if( c == 2 && (weight % 4 == 0 && weight !=0 && zeroflag != 0 || weight % 4 > 1) ) {
          result += two
        } else {
          result += number[c]
        }
      }
      result += unit(weight)
      zeroflag = 0
    }
    ++i
  }
  return result
}